---
title: 'Project 3: Variant calling in Burbot (*Lota lota*) using `freebayes` and `mpileup`'
author: "J Moggridge"
date: "06/03/2021"
output: 
  pdf_document:
    highlight: kate
    toc: TRUE
    toc_depth: 3
bibliography: references.bib
link-citations: true
---

```{r setup, echo=F}
knitr::opts_chunk$set(warning = F, message = F, eval = T, fig.align = 'center', echo = T)
```

# Introduction

Variant calling is a crucial step in population genetic analysis pipelines. Because high-throughput sequencing is inherently noisy, there is a need to apply stringent selection criteria for variants. How to decide on these criteria though, is neither simple nor easy to ascertain, as there exists no ground-truth for genomes. Indeed, there are many tools for variant calling that differ in their approaches (`bcftools`, `GATK`, `freebayes`, *etc.*). These generally have low concordance (cite stuff from hwang), but not in all cases and this may vary depending on the data in question [@hwang2015].

`bcftools` uses a two step process where `mpileup` first creates the 'pileup' of all reads at a position, and `call` then [@danecek2011, @li2011]

In this work I compared variant calls made by `bcftools` and `freebayes`. Both callers were used to obtain biallelic SNPs and results were compared in terms of the number of SNPs, the overlap between sets, the depth at each position, and their minor allele frequencies.

`freebayes` uses... @garrison2012

------------------------------------------------------------------------

# Methods

<!-- (chmod 755 /scratch/username/directory/ -R). -->

<!-- srun --pty --account="def-nricker" -t 0-03:00:00 --mem=32000 /bin/bash -->

### Dataset

I used data from project \#2 in this work, specifically the reads mapped to the burbot reference with the `bowtie2` aligner. I used `bcftools mpileup` and `GATK HaplotypeCaller` to call variants for the ten individuals using these alignments.

I first needed to convert my `.sam` files to `.bam` format and index them, as well as adding read groups using `picard` (necessary for `freebayes`).

```{r sam-to-bam and readgroups, engine = 'bash', eval = FALSE, echo=TRUE}
#!/bin/bash
# to convert .sam to .bam, then sort & index
module load samtools
for file in bam/*.sam; do
  name=`echo $file | sed 's/\.sam//'`
  samtools view -b -S -o $name.bam $file
  samtools sort $name.bam -o $name.sorted.bam;
  samtools index $name.sorted.bam
done

# add readgroups and re-index; show output to verify
module load picard 
group=1
for file in bam/*.sorted.bam
do
  label=`echo $file | sed 's/.sorted.bam//'`
  java -jar $EBROOTPICARD/picard.jar AddOrReplaceReadGroups I=$file \
    O=$label.rg.sorted.bam RGID=$group RGLB=lib1 RGPL=illumina RGPU=unit1 RGSM=$label
  samtools index $label.rg.sorted.bam
  echo "group number: $group; file: $file; read group:"
  samtools view -H $label.rg.sorted.bam | grep "@RG"
  group=$((group+1))
done
```

### `bcftools mpileup` and `freebayes` variant calling

I ran `bcftools` variant calling for the 10 .sorted.bam files. For `mpileup`, I added arguments to provide annotations for read depth and allelic depth (-a DP,AD), to skip indels, to specify Illumina as the sequencing platform (-P), to consider up to 1000 reads at a locus instead of the default 250 (--max-depth). For `call`, I specified to use the multi-allelic caller, to skip indels, and to provide the genotype quality scores (GQ).

```{r mpileup, engine = 'bash', eval = FALSE, echo=TRUE}
module load nixpkgs/16.09  gcc/7.3.0 bcftools/1.9
# mpileup and variant calling
bcftools mpileup -Ou -a DP,AD --skip-indels -P ILLUMINA  --max-depth 1000 \
-f ./burbot_ref/GCA_900302385.1_ASM90030238v1_genomic.fna ./bam/*.rg.sorted.bam | \
bcftools call -mv --format-fields GQ -o  mpileup.call.vcf
```

I called SNPs with `freebayes` with default settings and annotations for genotype quality. Indels were removed afterwards with `vcffilter`.

```{r freebayes, engine = 'bash', eval = FALSE, echo = TRUE}
module --force purge; module load nixpkgs/16.09  gcc/7.3.0 freebayes/1.2.0
# create a list of files to pass to as inputs
> files.txt
for file in bam/*.rg.sorted.bam; do echo $file >> files.txt; done
# free-bayes calling
freebayes -f burbot_ref/GCA_900302385.1_ASM90030238v1_genomic.fna  \
  --genotype-qualities --bam-list files.txt | \
vcffilter -f "TYPE = snp" > freebayes.call.vcf
```

### Filtering variant calls

Biallelic SNPs (`--min-alleles 2 --max-alleles 2 --remove-indels`) were retained if they had a minimum quality score of 30, minimum genotyping quality score of 10, and been genotyped in at least 50% of individuals (`--minQ 30 --minGQ 10 --max-missing 0.5`).

Given the small number of individuals, I did not feel it necessary to filter by allele frequency.

<!-- (--remove-filtered-all) is like  (--apply-filter "PASS")-->

<!-- (--min-alleles 2 --max-alleles 2) for biallelic -->

<!-- (--minQ 30 --minGQ 10 --max-missing 0.5) for QUAL, GQ, and prop missing data -->

<!-- (-recode --recode-INFO-all) to get a new vcf with annotation data -->

SHOULD ADD FILTER FOR: --maf 0.01 --max-maf 0.99

```{r filtering, engine='bash', eval = FALSE, echo = TRUE}
module --force purge; module load StdEnv/2020 vcftools/0.1.16
# filter mpileup calls 
vcftools --vcf mpileup.call.vcf --minQ 30 --minGQ 10 --max-missing 0.5 \
  --min-alleles 2 --max-alleles 2 --remove-indels \
  --remove-filtered-all --recode --recode-INFO-all --out mpileup.q30gq10g5
# same thing for freebayes
vcftools --vcf freebayes.call.vcf --minQ 30 --minGQ 10 --max-missing 0.5 \
  --min-alleles 2 --max-alleles 2 --remove-indels \
  --remove-filtered-all --recode --recode-INFO-all --out freebayes.q30gq10g5
```

There was no need to exclude any individuals from the analysis entirely, as none had more that 12% missing data (from `--missing-indv`).

```{r check_missing, engine='bash', eval = FALSE, echo = TRUE}
# check individuals' missing data
vcftools --vcf mpileup.q30gq10g5.recode.vcf --missing-indv \
  --out mpileup.q30gq10g5
cat mpileup.q30gq10g5.imiss
vcftools --vcf freebayes.q30gq10g5.recode.vcf --missing-indv \
  --out freebayes.q30gq10g5
cat freebayes.q30gq10g5.imiss
```

### Summarizing SNP calls

I counted and compared the biallelic SNP calls from each caller with `bcftools stats` and `isec`.

```{r count_snps, engine = 'bash', eval = FALSE, echo = TRUE}
module load StdEnv/2020 bcftools/1.10.2
# count the snps: 40,265 and 29,176
bcftools stats mpileup.q30gq10g5.recode.vcf | grep "number of SNPs:" | cut -f4
bcftools stats freebayes.q30gq10g5.recode.vcf | grep "number of SNPs:" | cut -f4 

# prepare mpileup snps by zipping and indexing
bgzip mpileup.q30gq10g5.recode.vcf
bcftools index mpileup.q30gq10g5.recode.vcf.gz
# Prepare freebayes snps by zipping and indexing
bgzip freebayes.q30gq10g5.recode.vcf
bcftools index freebayes.q30gq10g5.recode.vcf.gz

# Compute intersection of 2 files with isec
bcftools isec -p isec mpileup.q30gq10g5.recode.vcf.gz freebayes.q30gq10g5.recode.vcf.gz

# Get overlap data from isec into tsv format
> Overlap.vals
for file in isec/*.vcf
do 
  bcftools stats $file| grep "number of SNPs:" | cut -f4 >> Overlap.vals
done
echo $'unique to mpileup\nunique to freebayes\nshared\nshared' > Overlap.names
paste  Overlap.names Overlap.vals > Overlap.tsv
sed -i '1s/^/SNPs\tCount\n/' Overlap.tsv
```

Using `vcftools`:

-   Allele frequency at each loci across individuals

```{r vcf_freq, engine = 'bash', eval = FALSE, echo = TRUE}
vcftools --gzvcf mpileup.q30gq10g5.recode.vcf.gz --freq \
  --out mpileup.q30gq10g5
vcftools --gzvcf freebayes.q30gq10g5.recode.vcf.gz --freq \
  --out freebayes.q30gq10g5
```

-   Sum depth at each locus across individuals

```{r vcf_site_depth, engine = 'bash', eval = FALSE, echo = TRUE}
vcftools --gzvcf mpileup.q30gq10g5.recode.vcf.gz --site-depth \
  --out mpileup
vcftools --gzvcf freebayes.q30gq10g5.recode.vcf.gz --site-depth \
  --out freebayes
```

-   Mean depth for each individual across all loci, relatedness, and genotypes

```{r vcf_indv_depth, engine = 'bash', eval = FALSE, echo = TRUE}
# mean depth per individual
vcftools --gzvcf mpileup.q30gq10g5.recode.vcf.gz --depth --out mpileup
vcftools --gzvcf freebayes.q30gq10g5.recode.vcf.gz --depth --out freebayes

## Relatedness
vcftools --gzvcf mpileup.q30gq10g5.recode.vcf.gz --relatedness --out mpileup
vcftools --gzvcf freebayes.q30gq10g5.recode.vcf.gz --relatedness --out freebayes

## Convert to (0,1,2) format
vcftools --gzvcf mpileup.q30gq10g5.recode.vcf.gz --012 --out mpileup
vcftools --gzvcf freebayes.q30gq10g5.recode.vcf.gz --012 --out freebayes
```

<!-- quality, as shown in class, using what you think is a reasonable threshold. Compare the resulting vcf files in several ways. First, *compare the number of SNPs* and the *overlap of these SNPs across files*. -->

<!-- Then, plot the *minor allele frequency* and depth of all SNPs (summed across individuals), and discuss the variation across loci. Compare across vcfs produced by different variant callers. -->

    # get data down from cedar in local terminal:
    scp jmoggrid@cedar.computecanada.ca:/scratch/jmoggrid/Project3/Overlap.tsv .
    scp jmoggrid@cedar.computecanada.ca:/scratch/jmoggrid/Project3/*.frq .
    scp jmoggrid@cedar.computecanada.ca:/scratch/jmoggrid/Project3/*.idepth .
    scp jmoggrid@cedar.computecanada.ca:/scratch/jmoggrid/Project3/*.ldepth .

------------------------------------------------------------------------

### Statistics

I created figures and summary tables in the `R` language

```{r load data, message=FALSE}
#! (R code)
library(tidyverse, quietly = T)
library(ggbeeswarm)
library(patchwork)
library(rcartocolor)
library(gridExtra)
library(grid)

# Overlap counts
Overlap <- read_delim('./data/Overlap.tsv', '\t') %>%  
  distinct() %>% 
  bind_rows(
    tibble(SNPs = c('mpileup', 'freebayes'), 
           Count = c(.[1,2][[1]] + .[3,2][[1]], 
                     .[2,2][[1]] + .[3,2][[1]]))
    ) %>%  arrange(desc(Count))

# Allele frequencies
cnames <- c('chrom','pos', 'n_alleles','n_chr','a1','a2')
AlleleFreq <-  
  bind_rows(
    read_tsv('./data/freebayes.q30gq10g5.frq', col_names = cnames, skip = 1L) %>% 
      mutate(caller = 'freebayes'),
    read_tsv('./data/mpileup.q30gq10g5.frq', col_names = cnames, skip = 1L) %>% 
      mutate(caller = 'mpileup')
  ) %>% 
  mutate(allele = str_remove(a1, ':.+'), frequency = str_remove(a1, '.:'),
         allele2 = str_remove(a2, ':.+'), frequency2 = str_remove(a2, '.:')
  ) %>% 
  mutate(across(contains('freq'), as.numeric))

# Depth
cnames <- c('individual', 'n sites', 'mean_depth')
IndvDepth <- bind_rows(
    read_tsv('./data/mpileup.idepth', col_names = cnames, skip = 1L) %>%
      mutate(caller = 'mpileup'),
    read_tsv('./data/freebayes.idepth', col_names = cnames, skip = 1L) %>%
      mutate(caller = 'freebayes')
  ) %>% 
  mutate(individual = str_remove_all(individual, 'bam.|.bowtie.burbot'),
         caller = as_factor(caller)) %>% 
  arrange(individual)

# Site Depth
cnames <- c("chr", "pos", "allele_depth", "var_depth")
SiteDepth <- bind_rows(
  read_tsv('./data/mpileup.ldepth', col_names = cnames, skip = 1L) %>%
    mutate(caller = 'mpileup'),
  read_tsv('./data/freebayes.ldepth', col_names = cnames, skip = 1L) %>%
    mutate(caller = 'freebayes')
) %>% mutate(caller = as_factor(caller))

```

```{r figs, echo=F}
# Allele frequency histogram (alleles stacked)
fig_allelefreq <- 
  ggplot(AlleleFreq %>% filter(frequency != 0), 
         aes(x = frequency*10, fill=allele)) +
  geom_histogram(bins = 20, colour = 'white',) +
  facet_wrap(~caller) +
  rcartocolor::scale_fill_carto_d() +
  labs(x = 'minor allele frequency', y='loci', fill = 'Allele') +
  theme_bw()
# All allele freqs (including maf = 0, boo)
fig_allelefreq_all <- 
  ggplot(AlleleFreq, aes(x = G1.freq)) +
  geom_histogram(bins = 20, colour = 'white',) +
  facet_wrap(~caller) +
  scale_fill_carto_d() +
  labs(x = 'minor allele frequency', y='loci') +
  theme_bw()
# The distribution of depth per-site summed across individuals
fig_sitedepth <- 
  ggplot(SiteDepth, aes(x = allele_depth, fill = caller, color = caller)) +
  geom_density(alpha = 0.12) +
  geom_vline(aes(xintercept = mean(allele_depth), color = caller), lty = 2) +
  scale_x_log10() +
  labs(x = 'depth per locus', y='density') +
  guides(colour = guide_legend(override.aes = list(size=1, pch=15))) +
  theme_bw()
# The distribution of mean depth per site within individuals
set.seed(45)
fig_indvdepth <-
  ggplot(IndvDepth, aes(y = mean_depth, x = caller)) +
  geom_violin(alpha = 0.1, draw_quantiles = c(.25, .5, .75), trim = F) +
  geom_quasirandom(aes(color = individual), size = 3, width = 0.3) +
  scale_color_carto_d(palette = 2) +
  guides(colour = guide_legend(override.aes = list(size = 4, pch = 16))) +
  facet_wrap( ~ caller, scales = 'free_x') +
  labs(y = 'mean depth per locus', x = '', color = 'Individual') +
  theme_bw() +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    legend.position = 'right'
  )
```

------------------------------------------------------------------------

\newpage

# Results

Biallelic SNPs identified by each variant caller were counted and compared (table 1). `mpileup` called `r scales::comma(Overlap[1,2][[1]])` SNPs, `r round(Overlap[1,2][[1]] * 100 / Overlap[2,2][[1]], 2) - 100` % more than the `r scales::comma(Overlap[2,2][[1]])` called by `freebayes`. Of these, `r scales::comma(Overlap[3,2][[1]])` were shared between both sets, representing `r round(Overlap[3,2][[1]] * 100 /Overlap[1,2][[1]], 2)` % of `mpileup` calls and `r round(Overlap[3,2][[1]] * 100 /Overlap[2,2][[1]], 2)` % of `freebayes` calls. Additionally, `r scales::comma(Overlap[4,2][[1]])` SNPs were unique to `mpileup` and `r scales::comma(Overlap[5,2][[1]])` to `freebayes`.

The allele frequency....

The sum depth per locus across all individuals is generally in the range \~ 50-200 reads, with both callers showing similar distributions (fig 1.C). The mean depth for `mpileup` was 173.7 reads and for `freebayes` 162.2 reads. Both callers have some outliers in terms of sum depth across individuals: the maximum depth for `freebayes` was 56,564 reads, while for `freebayes` it was 10,046 reads (as the max-depth consideration of 1000 was applied here).

------------------------------------------------------------------------

```{r fig1, echo=F, fig.align = 'center', fig.width=9, fig.height=5.5, fig.cap="(A) Distribution of depth per locus summed across all individuals (note log-scaled x-axis). (B) Mean depth per locus for each individual. (C) Distribution of minor allele frequency; SNPs with frequency = 0 or 1 were omitted. (D)Total, shared, and unique SNPs called by mpileup and freebayes"}
(
  (fig_allelefreq + fig_indvdepth) /
  (fig_sitedepth + gridExtra::tableGrob(Overlap, rows = NULL))
) &
  plot_layout(guides = 'keep') &
  plot_annotation(tag_levels = 'A')
```

\newpage

# Discussion

------------------------------------------------------------------------

\newpage

# References
